src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
+            if (-1 != referenceIndex) {
+                setReferenceIndex(referenceIndex); // sets reference name and index
             }
+            else {
+                mReferenceName = referenceName.intern();
+                mReferenceIndex = null;
+            }
+        }
+        else {
+            mReferenceName = referenceName.intern();
+            mReferenceIndex = null;
         }
-        // Drop through from above if nothing done.
-        mReferenceName = value.intern();
-        mReferenceIndex = null;
     }
 
     /**
-     * @return index of the reference sequence for this read in the sequence dictionary, or -1
-     * if read has no reference sequence set, or if a String reference name is not found in the sequence index..
+     * Returns the reference index for this record.
+     *
+     * If the reference name for this record has previously been resolved against the sequence dictionary, the corresponding
+     * index is returned directly. Otherwise, the record must have a non-null SAMFileHeader that can be used to
+     * resolve the index for the record's current reference name. If the record has a header, and the name does not appear
+     * in the header's sequence dictionary, the value NO_ALIGNMENT_REFERENCE_INDEX (-1) will be returned. If the record
+     * does not have a header, an IllegalArgumentException is thrown.
+     *
+     * @return Index in the sequence dictionary of the reference sequence. If the read has no reference sequence, or if
+     * the reference name is not found in the sequence index, NO_ALIGNMENT_REFERENCE_INDEX (-1) is returned.
+     *
+     * @throws IllegalArgumentException if the reference index cannot be resolved because the SAMFileHeader for the
+     * record is null.
      */
     public Integer getReferenceIndex() {
-        if (mReferenceIndex == null) {
-            if (mReferenceName == null) {
-                mReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
-            } else if (NO_ALIGNMENT_REFERENCE_NAME.equals(mReferenceName)) {
+        if (null == mReferenceIndex) {
+            // try to resolve the reference index
+            if (NO_ALIGNMENT_REFERENCE_NAME.equals(mReferenceName)) {
                 mReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
-            } else {
+            }
+            else if (null != mHeader) {
                 mReferenceIndex = mHeader.getSequenceIndex(mReferenceName);
             }
+            else {
+                throw new IllegalArgumentException("A non-null SAMFileHeader is required to resolve the reference index");
+            }
         }
         return mReferenceIndex;
     }
 
     /**
-     * @param referenceIndex Must either equal -1 (indicating no reference), or exist in the sequence dictionary
-     * in the header associated with this record.
+     * Updates the reference index. The record must have a valid SAMFileHeader, and the reference index must appear in
```

Comment:
"The record must have a valid SAMFileHeader, unless the referenceIndex parameter equals NO_ALIGNMENT_REFERENCE_INDEX"